pkg/api/controller.go

```diff
@@ -99,6 +99,225 @@ type Controller struct {
 	otfDiffService        *tablediff.Service
 }
 
+func (c *Controller) CreatePresignMultipartUpload(w http.ResponseWriter, r *http.Request, repository string, branch string, params apigen.CreatePresignMultipartUploadParams) {
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.WriteObjectAction,
+			Resource: permissions.ObjectArn(repository, params.Path),
+		},
+	}) {
+		return
+	}
+	ctx := r.Context()
+	c.LogAction(ctx, "create_presign_multipart_upload", r, repository, branch, "")
+
+	// check if api is supported
+	storageConfig := c.getStorageConfig()
+	if !swag.BoolValue(storageConfig.PreSignMultipartUpload) {
+		writeError(w, r, http.StatusNotImplemented, "presign multipart upload API is not supported")
+		return
+	}
+
+	repo, err := c.Catalog.GetRepository(ctx, repository)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	// check if the branch exists - it is still a possibility, but we don't want to
+	// upload to start and fail at the end when the branch was not there in the first place
+	branchExists, err := c.Catalog.BranchExists(ctx, repository, branch)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+	if !branchExists {
+		writeError(w, r, http.StatusNotFound, fmt.Sprintf("branch '%s' not found", branch))
+		return
+	}
+
+	// generate a new address for the object we like to upload
+	address := c.PathProvider.NewPath()
+	qk, err := c.BlockAdapter.ResolveNamespace(repo.StorageNamespace, address, block.IdentifierTypeRelative)
+	if err != nil {
+		writeError(w, r, http.StatusInternalServerError, err)
+		return
+	}
+	err = c.Catalog.SetLinkAddress(ctx, repository, address)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	// create a new multipart upload
+	mpuResp, err := c.BlockAdapter.CreateMultiPartUpload(ctx, block.ObjectPointer{
+		StorageNamespace: repo.StorageNamespace,
+		IdentifierType:   block.IdentifierTypeRelative,
+		Identifier:       address,
+	}, nil, block.CreateMultiPartUploadOpts{})
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	// prepare presigned URL, for each part
+	var presignedURLs []string
+	for i := 0; i < swag.IntValue(params.Parts); i++ {
+		// generate a pre-signed PUT url for the given request
+		preSignedURL, err := c.BlockAdapter.GetPresignUploadPartURL(ctx, block.ObjectPointer{
+			StorageNamespace: repo.StorageNamespace,
+			Identifier:       address,
+			IdentifierType:   block.IdentifierTypeRelative,
+		}, mpuResp.UploadID, i+1)
+		if c.handleAPIError(ctx, w, r, err) {
+			return
+		}
+		presignedURLs = append(presignedURLs, preSignedURL)
+	}
```

Comment:
Can this not work effectively for large parts amount?

Reply:
Yes, my two main concerns are:
- Large files which require long time to upload which make the URLs expire. I assume it can be the same case for a single presign URL but I am not sure AWS will expire the link during a single upload.
- Large number of parts that will make the response, ex: 10k lines 1k len each gets a response of ~10mb.

Adding another API for pulling pool of URLs for mpu will help in these cases.
I'll open an issue to describe the streaming and the above use case and associate it here.